Lwt 4.3 breaks compilation with musl

To reproduce:

```
% opam switch create 4.09.0+musl+flambda
% cat test.ml
let () = Lwt_main.run (Lwt.return ())
% cat dune
(executable (name test) (libraries lwt lwt.unix))
% dune build test.exe
    ocamlopt test.exe (exit 2)
(cd _build/default && ~/.opam/4.09.0+musl+flambda/bin/ocamlopt.opt -w @1..3@5..28@30..39@43@46..47@49..57@61..62-40 -strict-sequence -strict-formats -short-paths -keep-locs -g -o test.exe -I ~/.opam/4.09.0+musl+flambda/lib/bytes -I ~/.opam/4.09.0+musl+flambda/lib/lwt -I ~/.opam/4.09.0+musl+flambda/lib/lwt/unix -I ~/.opam/4.09.0+musl+flambda/lib/mmap -I ~/.opam/4.09.0+musl+flambda/lib/ocaml/threads -I ~/.opam/4.09.0+musl+flambda/lib/ocplib-endian -I ~/.opam/4.09.0+musl+flambda/lib/result -I ~/.opam/4.09.0+musl+flambda/lib/seq ~/.opam/4.09.0+musl+flambda/lib/result/result.cmxa ~/.opam/4.09.0+musl+flambda/lib/lwt/lwt.cmxa ~/.opam/4.09.0+musl+flambda/lib/ocaml/unix.cmxa ~/.opam/4.09.0+musl+flambda/lib/ocaml/bigarray.cmxa ~/.opam/4.09.0+musl+flambda/lib/mmap/mmap.cmxa ~/.opam/4.09.0+musl+flambda/lib/ocplib-endian/bigstring.cmxa ~/.opam/4.09.0+musl+flambda/lib/ocaml/threads/threads.cmxa ~/.opam/4.09.0+musl+flambda/lib/lwt/unix/lwt_unix.cmxa .test.eobjs/native/test.cmx)
/opt/x86_64-linux-musl-native/bin/../lib/gcc/x86_64-linux-musl/9.1.0/../../../../x86_64-linux-musl/bin/ld: ~/.opam/4.09.0+musl+flambda/lib/ocaml/libthreadsnat.a(st_stubs_n.o): in function `caml_thread_start':
st_stubs.c:(.text+0xab1): undefined reference to `sigsetjmp'
/opt/x86_64-linux-musl-native/bin/../lib/gcc/x86_64-linux-musl/9.1.0/../../../../x86_64-linux-musl/bin/ld: ~/.opam/4.09.0+musl+flambda/lib/ocaml/libasmrun.a(startup_nat_n.o): in function `caml_startup_common':
~/.opam/4.09.0+musl+flambda/.opam-switch/build/ocaml-variants.4.09.0+musl+flambda/runtime/startup_nat.c:156: undefined reference to `sigsetjmp'
/opt/x86_64-linux-musl-native/bin/../lib/gcc/x86_64-linux-musl/9.1.0/../../../../x86_64-linux-musl/bin/ld: ~/.opam/4.09.0+musl+flambda/lib/ocaml/libasmrun.a(unix_n.o): in function `caml_secure_getenv':
~/.opam/4.09.0+musl+flambda/.opam-switch/build/ocaml-variants.4.09.0+musl+flambda/runtime/unix.c:418: undefined reference to `issetugid'
/opt/x86_64-linux-musl-native/bin/../lib/gcc/x86_64-linux-musl/9.1.0/../../../../x86_64-linux-musl/bin/ld: ~/.opam/4.09.0+musl+flambda/lib/ocaml/libasmrun.a(unix_n.o): in function `caml_dlopen':~/.opam/4.09.0+musl+flambda/.opam-switch/build/ocaml-variants.4.09.0+musl+flambda/runtime/unix.c:278: undefined reference to `dlopen'/opt/x86_64-linux-musl-native/bin/../lib/gcc/x86_64-linux-musl/9.1.0/../../../../x86_64-linux-musl/bin/ld: ~/.opam/4.09.0+musl+flambda/lib/ocaml/libasmrun.a(unix_n.o): in function `caml_dlclose':~/.opam/4.09.0+musl+flambda/.opam-switch/build/ocaml-variants.4.09.0+musl+flambda/runtime/unix.c:284: undefined reference to `dlclose'
/opt/x86_64-linux-musl-native/bin/../lib/gcc/x86_64-linux-musl/9.1.0/../../../../x86_64-linux-musl/bin/ld: ~/.opam/4.09.0+musl+flambda/lib/ocaml/libasmrun.a(unix_n.o): in function `caml_dlsym':
~/.opam/4.09.0+musl+flambda/.opam-switch/build/ocaml-variants.4.09.0+musl+flambda/runtime/unix.c:289: undefined reference to `dlsym'
/opt/x86_64-linux-musl-native/bin/../lib/gcc/x86_64-linux-musl/9.1.0/../../../../x86_64-linux-musl/bin/ld: ~/.opam/4.09.0+musl+flambda/.opam-switch/build/ocaml-variants.4.09.0+musl+flambda/runtime/unix.c:289: undefined reference to `dlsym'
/opt/x86_64-linux-musl-native/bin/../lib/gcc/x86_64-linux-musl/9.1.0/../../../../x86_64-linux-musl/bin/ld: ~/.opam/4.09.0+musl+flambda/lib/ocaml/libasmrun.a(unix_n.o): in function `caml_dlerror':~/.opam/4.09.0+musl+flambda/.opam-switch/build/ocaml-variants.4.09.0+musl+flambda/runtime/unix.c:303: undefined reference to `dlerror'
collect2: error: ld returned 1 exit status
File "caml_startup", line 1:Error: Error during linking
```

Looks like `lwt_unix.cmxa` adds `/usr/lib` to the search path:

```
% ocamlobjinfo ~/.opam/4.09.0+musl+flambda/lib/lwt/unix/lwt_unix.cmxa|grep object
Extra C object files: -llwt_unix_stubs -L/usr/lib -lpthread
```

In 4.2.1 it works fine:

```
% opam install lwt=4.2.1
The following actions will be performed:
⊘ removed   lwt.4.3.0
∗ installed lwt.4.2.1
Done.
% dune build test.exe
% ocamlobjinfo ~/.opam/4.09.0+musl+flambda/lib/lwt/unix/lwt_unix.cmxa|grep object
Extra C object files: -llwt_unix_stubs -lpthread 
```

This switch uses musl-gcc, which uses its own search path. It isn't compatible with objects from `/usr/lib` which were compiled with glibc. 

<hr>

Relatedly, when installing conf-libev, I need to pass `C_INCLUDE_PATH` and `LIBRARY_PATH` even though my compiler already has the right search paths baked in:

```
% cat test.c
#include <ev.h>
int main() {}
% musl-gcc test.c -lev
```

Where is `ev.h` installed on your system, or for musl-gcc?

I get:

```
$ musl-gcc test.c -lev
test.c:1:16: fatal error: ev.h: No such file or directory
compilation terminated.
```

This is under Ubuntu, with no libev installed specifically for musl-gcc to use, so it's what I'd expect.

--- a/.travis.yml
+++ b/.travis.yml
@@ -7,7 +7,7 @@ matrix:
     - os: linux
       env: COMPILER=4.09.0 LIBEV=no PPX_LET=yes COVERAGE=yes
     - os: linux
-      env: COMPILER=ocaml-variants.4.08.1+flambda
+      env: COMPILER=4.08.1+musl+flambda MUSL=yes
     - os: linux
       env: COMPILER=4.07.1 DOCS=yes
     - os: linux
@@ -22,7 +22,7 @@ matrix:
       env: COMPILER=4.02.3 PACKAGING=yes
 
   allow_failures:
-    - env: COMPILER=ocaml-variants.4.08.1+flambda
+    - env: COMPILER=4.08.1+musl+flambda MUSL=yes
     - env: COMPILER=4.07.1 DOCS=yes
     - env: COMPILER=4.06.1 LWT_STRESS_TEST=true
     - env: COMPILER=4.05.0
@@ -39,8 +39,26 @@ before_install:
   - '[ "$TRAVIS_EVENT_TYPE" != cron ] || rm -rf ~/.opam ./_opam ./_cache'
 
 install:
-  - '[ "$LIBEV" == no ] || sudo apt-get update -qq'
-  - '[ "$LIBEV" == no ] || sudo apt-get install -qq libev-dev'
+  - |
+    if [ "$LIBEV" != no ]
+    then
+      sudo apt-get update -qq
+      if [ "$MUSL" != yes ]
+      then
+        sudo apt-get install -qq libev-dev
+      else
+        sudo apt-get install musl-tools
+        wget http://dist.schmorp.de/libev/Attic/libev-4.27.tar.gz
+        tar xf libev-4.27.tar.gz
+        cd libev-4.27
+        CC=musl-gcc ./configure --prefix `pwd`/install
+        make
+        make install
+        sudo cp install/include/* /usr/include/x86_64-linux-musl
+        sudo cp install/lib/* /usr/lib/x86_64-linux-musl
+        cd ..
+      fi
+    fi
 
   - OS=linux
   - '[ "$TRAVIS_OS_NAME" != osx ] || OS=macos'
@@ -59,6 +77,7 @@ install:
   - opam --version
   - ocaml -version
 
+  - opam pin add --no-action ounit 2.0.8 # 2.1.2 release is broken.
   - CACHED=yes
   - '[ -d _cache/_build ] || CACHED=no'
   - '[ "$CACHED" == yes ] || opam install . --deps-only --yes'

--- a/src/unix/config/discover.ml
+++ b/src/unix/config/discover.ml
@@ -439,9 +439,7 @@ struct
         None
       else begin
         skip_if_windows context @@ fun () ->
-        if !Arguments.android_target <> Some true then
-          C_library_flags.detect context ~library:"pthread";
-        compiles context {|
+        let code = {|
           #include <pthread.h>
 
           int main()
@@ -450,6 +448,16 @@ struct
               return 0;
           }
         |}
+        in
+        match compiles context code with
+        | Some true -> Some true
+        | no ->
+          if !Arguments.android_target = Some true then
+            no
+          else begin
+            C_library_flags.detect context ~library:"pthread";
+            compiles context code
+          end
       end
   }
 

--- a/src/unix/config/discover.ml
+++ b/src/unix/config/discover.ml
@@ -156,6 +156,7 @@ sig
 
   val c_flags : unit -> string list
   val link_flags : unit -> string list
+  val add_link_flags : string list -> unit
 end =
 struct
   let c_flags = ref []
@@ -164,6 +165,9 @@ struct
   let extend c_flags' link_flags' =
     c_flags := !c_flags @ c_flags';
     link_flags := !link_flags @ link_flags'
+
+  let add_link_flags flags =
+    extend [] flags
 
   let (//) = Filename.concat
 
@@ -357,7 +361,7 @@ struct
         Output.{name = feature.macro_name; found}
     end
 
-  let compiles ?(werror = false) context code =
+  let compiles ?(werror = false) ?(link_flags = []) context code =
     let c_flags = C_library_flags.c_flags () in
     let c_flags =
       if werror then
@@ -365,8 +369,8 @@ struct
       else
         c_flags
     in
-    Configurator.c_test
-      context ~c_flags ~link_flags:(C_library_flags.link_flags ()) code
+    let link_flags = link_flags @ (C_library_flags.link_flags ()) in
+    Configurator.c_test context ~c_flags ~link_flags code
     |> fun result -> Some result
 
   let skip_if_windows context k =
@@ -418,8 +422,7 @@ struct
       if not should_look_for_libev then
         None
       else begin
-        C_library_flags.detect context ~library:"ev";
-        compiles context {|
+        let code = {|
           #include <ev.h>
 
           int main()
@@ -428,6 +431,14 @@ struct
               return 0;
           }
         |}
+        in
+        match compiles context code ~link_flags:["-lev"] with
+        | Some true ->
+          C_library_flags.add_link_flags ["-lev"];
+          Some true
+        | _ ->
+          C_library_flags.detect context ~library:"ev";
+          compiles context code
       end
   }
 

--- a/test/test.ml
+++ b/test/test.ml
@@ -325,7 +325,7 @@ let concurrent library_name suites =
     flush stdout;
     outcomes |> List.iter (function
       | (suite, test), Failed ->
-        Printf.eprintf "Test '%s' in suite '%s' produced 'false\n"
+        Printf.eprintf "Test '%s' in suite '%s' produced 'false'\n"
           test.test_name suite.suite_name
       | (suite, test), Exception exn ->
         Printf.eprintf "Test '%s' in suite '%s' raised '%s'\n"

--- a/test/unix/test_lwt_io.ml
+++ b/test/unix/test_lwt_io.ml
@@ -497,14 +497,16 @@ let suite = suite "lwt_io" [
     Lwt_bytes.of_string "\x70\x60\x50\x40\x30\x20\xf0\x42"
     |> Lwt_io.(of_bytes ~mode:input)
     |> Lwt_io.LE.read_float64
-    >|= (=) 283686952306183.
+    >|= Int64.bits_of_float
+    >|= (=) 0x42F0203040506070L
   end;
 
   test "NumberIO.BE.read_float64" begin fun () ->
     Lwt_bytes.of_string "\x42\xf0\x20\x30\x40\x50\x60\x70"
     |> Lwt_io.(of_bytes ~mode:input)
     |> Lwt_io.BE.read_float64
-    >|= (=) 283686952306183.
+    >|= Int64.bits_of_float
+    >|= (=) 0x42F0203040506070L
   end;
 
   test "NumberIO.LE.write_int" begin fun () ->
@@ -580,14 +582,14 @@ let suite = suite "lwt_io" [
   test "NumberIO.LE.write_float64" begin fun () ->
     let buffer = Lwt_bytes.create 8 in
     Lwt_io.LE.write_float64 (Lwt_io.(of_bytes ~mode:output) buffer)
-      283686952306183. >>= fun () ->
+      (Int64.float_of_bits 0x42F0203040506070L) >>= fun () ->
     Lwt.return (Lwt_bytes.to_string buffer = "\x70\x60\x50\x40\x30\x20\xf0\x42")
   end;
 
   test "NumberIO.BE.write_float64" begin fun () ->
     let buffer = Lwt_bytes.create 8 in
     Lwt_io.BE.write_float64 (Lwt_io.(of_bytes ~mode:output) buffer)
-      283686952306183. >>= fun () ->
+      (Int64.float_of_bits 0x42F0203040506070L) >>= fun () ->
     Lwt.return (Lwt_bytes.to_string buffer = "\x42\xf0\x20\x30\x40\x50\x60\x70")
   end;
 ]

